feat: implementation of whitebox manipulator

pollend opened the pull request:
implement manipulators for TransformMode. At the moment I'll just limit it to manipulate polygon faces. There are some things that can be added to improve this more but I'll just limit this change to just the faces to just get the basic functionality there. 

## Translation

![Peek 2022-06-12 13-31](https://user-images.githubusercontent.com/854359/173252536-3d667681-845e-4d10-998c-1237c3449757.gif)

## Rotation

https://user-images.githubusercontent.com/854359/173252642-e6bc1dcc-238d-4ed7-b2cd-103e69fa9524.mp4

## Scale

https://user-images.githubusercontent.com/854359/173252640-65ef0d33-2f7f-40a2-ab82-023ebe10e3d3.mp4




At the moment the selection is only limited to selecting faces only but I think can be be expanded to allow manipulating faces/edges/vertices. At the moment all the manipulation are in local space  I can see this working in normal space i.e the normal of the selected polygon or the average if there are multiple. I can also see view space i.e from the perspective of the camera. 

Signed-off-by: Michael Pollind <mpollind@gmail.com>

hultonha commented on Gems/WhiteBox/Code/Source/EditorWhiteBoxComponentMode.cpp:
```diff
@@ -538,8 +545,8 @@ namespace WhiteBox
         {
-            bool hasTransformMode = false;
-            settingsRegistry->Get(hasTransformMode, WhiteBoxTransformFeature);
+            bool hasTransformMode = true;
+            // settingsRegistry->Get(hasTransformMode, WhiteBoxTransformFeature);
```
update: Would be better to just add this setting to your setreg file if possible?

pollend replied:
I should do that, not sure why I did this because I have it already in the json.

hultonha commented on Gems/WhiteBox/Code/Source/EditorWhiteBoxComponentMode.h:
```diff
@@ -30,6 +30,14 @@ namespace WhiteBox
 
+    enum class TransformType 
+    {
+        TranslationType,
+        RotationType,
+        ScaleType,
```
nit: omit `Type` postfix as `TransformType` has `Type` in the enum name

hultonha commented on Gems/WhiteBox/Code/Source/SubComponentModes/EditorWhiteBoxTransformMode.cpp:
```diff
@@ -52,4 +66,32 @@ namespace WhiteBox
+
+    void TransformMode::CreateRotationManipulators()
+    {
+        // AZStd::unique_ptr<AzToolsFramework::RotationManipulators> translationManipulators = AZStd::make_unique<AzToolsFramework::RotationManipulators>(
+        //     AzToolsFramework::TranslationManipulators::Dimensions::Three, AZ::Transform::CreateIdentity(), AZ::Vector3::CreateOne());
+        // translationManipulators->Register(AzToolsFramework::g_mainManipulatorManagerId);
+        
+        // m_manipulator = AZStd::move(translationManipulators);
```
update: I believe you need to use a `shared_ptr` for manipulators to help deal with lifetime issues (this is explicit in the interface to all manipulators (see `MakeShared`)

hultonha commented on Gems/WhiteBox/Code/Source/Viewport/WhiteBoxPolygonGizmoTranslation.h:
```diff
@@ -0,0 +1,26 @@
+#include <Manipulators/TranslationManipulators.h>
+#include <AzToolsFramework/Viewport/ViewportTypes.h>
+
+namespace WhiteBox
+{
+    class PolygonGizmoTranslation : public AzToolsFramework::TranslationManipulators {
+        public:
+            void ProcessManipulators(const AZStd::function<void(BaseManipulator*)>&) override;
+
+        private:
+            MeshSelection m_selection_manipulator;
+            // ManipulatorViewPolygon m_polygonView;
+            // AzToolsFramework::ManipulatorViewSphere m_vertexView; //!< Manipulator view used to represent a mesh vertex for selection.
+    };
+}
```
update: I would not add this, instead, put all the logic in callbacks for the individual manipulators, no need to create a new type an inherit

Have a look at `EditorTransformComponentSelection::CreateTranslationManipulators()` for an example

```
// in EditorWhiteBoxTransformMode

translationManipulators->InstallLinearManipulatorMouseDownCallback(
    [this]([[maybe_unused]] const LinearManipulator::Action& action)
    {
        // do White Box logic here, use `action` which will be how the mouse moved the manipulator
    }
```

hultonha commented on Gems/WhiteBox/Code/Source/Viewport/WhiteboxMeshSelectionManipulator.h:
```diff
@@ -0,0 +1,10 @@
+#include "Manipulators/ManipulatorBus.h"
+
+
+namespace WhiteBox
+{
+class MeshSelectionManipulator : public AzToolsFramework::BaseManipulator, 
+    public AzToolsFramework::ManipulatorSpace{
+
+};
+}
```
update: This can be removed too, much better to use the selection primitives provided by EditorWhiteBoxComponentMode (e.g. `FindClosestVertexIntersection`, `FindClosestEdgeIntersection` etc... you can use that returned edge to create a Manipulator when one of these is returned. This is what we do in DefaultMode right now. For this new transform mode, we would just want to create a `TranslationManipulator` that displays until you click another vertex or click off of the manipulator (similar to how the Spline works with EditorVertexSelection)

pollend replied:
yep, that's what I was planning to do as well.

hultonha reviewed (commented):
Added a few suggestions, please give me a shout if you have any other questions :+1:

pollend commented on Gems/WhiteBox/Code/Source/SubComponentModes/EditorWhiteBoxTransformMode.h:
```diff
@@ -52,12 +68,33 @@ namespace WhiteBox
+        void CreateRotationManipulators();
+        void CreateScaleManipulators();
+
+        PolygonSelection refreshPolygonSelection(WhiteBoxMesh* whitebox, const PolygonIntersection& intersection);
```
for the moment I'm just going to use this to update the selection. might make sense to move this into a class maybe?

hultonha replied:
Okay sounds good, let me know how you get on 👍

pollend commented on Gems/WhiteBox/Code/Source/Util/WhiteBoxMathUtil.cpp:
```diff
@@ -94,4 +94,20 @@ namespace WhiteBox
     }
+
+
+    AZStd::vector<AZ::Vector3> TransformToWorldSpace(
```
I probably can revert this change I ended up not using it.

pollend commented on Gems/WhiteBox/Code/Source/SubComponentModes/EditorWhiteBoxTransformMode.cpp:
```diff
@@ -18,38 +35,457 @@
+
+        if(m_polygonIntersection.has_value()) {
+            auto& polygonIntersection = m_polygonIntersection.value();
+            DrawFace(debugDisplay, whiteBox, polygonIntersection.GetHandle(), cl_whiteBoxPolygonHoveredColor);
+            DrawOutline(debugDisplay, whiteBox, polygonIntersection.GetHandle(), cl_whiteBoxPolygonHoveredOutlineColor);
```
I don't really like how these properties are by manipulator action. could we define a set of fixed colors for selection/selected for the mesh outline, vertices and edges. I don't imagine the colors should vary if much at all between all these modes.

hultonha reviewed (commented):
This looks like an absolutely incredible start @pollend, amazing work! And thanks for including the excellent videos too! This looks genuinely awesome 😄 

I have a fair amount of feedback but it's very much heading in the right direction. Tagging @greerdv for visibility too as they've been working a lot with manipulators recently and may have some pointers/ideas/suggestions 🙂 

Keep up the awesome work and I'll re-review this when I'm back in a little over a week 👍

pollend commented on Gems/WhiteBox/Code/Source/EditorWhiteBoxComponentMode.cpp:
```diff
@@ -87,9 +95,32 @@ namespace WhiteBox
             break;
         }
+        
+        switch (m_transformType)
+        {
+        case WhiteBox::TransformType::Translation:
+            SetViewportUiClusterActiveButton(m_transformClusterId, m_transformTranslateButtonId);
+            break;
+        case WhiteBox::TransformType::Rotation:
+            SetViewportUiClusterActiveButton(m_transformClusterId, m_transformRotateButtonId);
+            break;
+        case WhiteBox::TransformType::Scale:
+            SetViewportUiClusterActiveButton(m_transformClusterId, m_transformScaleButtonId);
+            break;
+        default:
+            break;
+        }
```
Maybe just make the cluster on demand. I could place it on the right side of the screen. If it's on the left it can be placed after the mode selection. I can just move all of this into the mode itself

hultonha replied:
Could be a question for sig/ui-ux - it should be possible to stack the cluster to the right of the existing one too - that would be great thanks!

pollend commented on Gems/WhiteBox/Code/Source/SubComponentModes/EditorWhiteBoxTransformMode.cpp:
```diff
@@ -6,50 +6,497 @@
         return {};
     }
 
+    void TransformMode::DrawFace(
+        AzFramework::DebugDisplayRequests& debugDisplay, WhiteBoxMesh* mesh, const Api::PolygonHandle& polygon, const AZ::Color& color)
+    {
+        AZStd::vector<AZ::Vector3> triangles = Api::PolygonFacesPositions(*mesh, polygon);
+        const AZ::Vector3 normal = Api::PolygonNormal(*mesh, polygon);
+
+        if (!triangles.empty())
+        {
+            // draw selected polygon
+            debugDisplay.PushMatrix(AZ::Transform::CreateTranslation(normal * ed_whiteBoxPolygonViewOverlapOffset));
+
+            debugDisplay.SetColor(color);
+            debugDisplay.DrawTriangles(triangles, color);
+
+            debugDisplay.PopMatrix();
+        }
+    }
```
not at the moment its been implemented for the WhiteBoxManipulatorViews.cpp

pollend commented on Gems/WhiteBox/Code/Source/SubComponentModes/EditorWhiteBoxTransformMode.cpp:
```diff
@@ -6,50 +6,497 @@
+
+        auto mouseMoveHandlerFn =
+            [entityComponentIdPair = m_entityComponentIdPair, sharedState,
+            translationManipulator = AZStd::weak_ptr<AzToolsFramework::TranslationManipulators>(translationManipulators)](const auto& action)
```
seems to work correctly umm its probably not what is intended though ...

https://user-images.githubusercontent.com/854359/174525555-0c017558-0d37-41c7-8eda-bb11c1b8b411.mp4

pollend commented:
I ended up overhauling this to some degree. 

https://user-images.githubusercontent.com/854359/174527651-f20f8090-af3e-49a0-bb02-0d36c2b4ef9d.mp4

hultonha commented:
Building locally I got a compile error for this usage

```
Severity	Code	Description	Project	File	Line	Suppression State
Error	C5233	explicit lambda capture 'scaleManipulator' is not used
             scaleManipulator = AZStd::weak_ptr<AzToolsFramework::ScaleManipulators>(scaleManipulators)](const auto& action)
             ^	WhiteBox.Editor.Static	D:\o3de\Gems\WhiteBox\Code\Source\SubComponentModes\EditorWhiteBoxTransformMode.cpp	537	
```

pollend commented:
> Building locally I got a compile error for this usage
> 
> ```
> Error	C5233	explicit lambda capture 'scaleManipulator' is not used
>              scaleManipulator = AZStd::weak_ptr<AzToolsFramework::ScaleManipulators>(scaleManipulators)](const auto& action)
>              ^	WhiteBox.Editor.Static	D:\o3de\Gems\WhiteBox\Code\Source\SubComponentModes\EditorWhiteBoxTransformMode.cpp	537	
> ```

Is it an actual problem. Clang doesn't seem to catch this.

hultonha commented:
> > Building locally I got a compile error for this usage
> > ```
> > Error	C5233	explicit lambda capture 'scaleManipulator' is not used
> >              scaleManipulator = AZStd::weak_ptr<AzToolsFramework::ScaleManipulators>(scaleManipulators)](const auto& action)
> >              ^	WhiteBox.Editor.Static	D:\o3de\Gems\WhiteBox\Code\Source\SubComponentModes\EditorWhiteBoxTransformMode.cpp	537	
> > ```
> 
> Is it an actual problem. Clang doesn't seem to catch this.

It is yeah, unused variable warning, there was another instance on line 559 too (there's another with `renderData` in `Display` too I'm afraid)

hultonha commented on Gems/WhiteBox/Code/Source/Viewport/WhiteBoxViewportConstants.h:
```diff
@@ -29,6 +29,14 @@ namespace WhiteBox
     AZ_CVAR_EXTERNED(float, cl_whiteBoxEdgeSelectionWidth);
     AZ_CVAR_EXTERNED(float, cl_whiteBoxSelectedEdgeVisualWidth);
+    
+    AZ_CVAR_EXTERNED(AZ::Color, ed_whiteBoxVertexHover);
+    AZ_CVAR_EXTERNED(AZ::Color, ed_whiteBoxPolygonHover);
+    AZ_CVAR_EXTERNED(AZ::Color, ed_whiteBoxOutlineHover);
+    AZ_CVAR_EXTERNED(AZ::Color, ed_whiteBoxPolygonSelection);
+    AZ_CVAR_EXTERNED(AZ::Color, ed_whiteBoxVertexSelection);
+    AZ_CVAR_EXTERNED(AZ::Color, ed_whiteBoxOutlineSelection);
+    AZ_CVAR_EXTERNED(float, ed_whiteBoxPolygonViewOverlapOffset);
```
update: See my earlier comment about the other existing draw functions - we might want to try unify them all so we don't get drifting functionality and have a consistent function to use. If you could take a little look at this that'd be great. Thanks!

pollend replied:
the plan was to describe the action and just update everything to follow that pattern. avoid having to define colors for everything. would it be better to define it for everything or for the use case :?

pollend replied:
here is a PR that I prepared for this: https://github.com/o3de/o3de/pull/10510

hultonha reviewed (commented):
Updates are looking great, there's just a few little things to follow-up on and then we should get this merged. Awesome work @pollend 😄

pollend commented:
@hultonha for `InstallSurfaceManipulatorMouseMoveCallback` thats where the mouse sticks to a surface right? It does work just not sure how this should work functionally. anything you have in mind that I should tweak about this?

hultonha commented on Gems/WhiteBox/Code/Source/Util/WhiteBoxEditorDrawUtil.cpp:
```diff
@@ -0,0 +1,80 @@
+
+namespace WhiteBox
+{
+    void DrawFace(
```
That's fair, but yeah a fast-follow PR to unify these would be great

hultonha commented:
> @hultonha for `InstallSurfaceManipulatorMouseMoveCallback` thats where the mouse sticks to a surface right? It does work just not sure how this should work functionally. anything you have in mind that I should tweak about this? if you can get this working in your workspace you can probably point out some things to improve there and we can work out that behavior.

Cool I'll give this a try locally to see if anything comes up, might be fine but I'd like to just double check it :+1:

hultonha commented:
Hey @pollend, I just tried building this again after grabbing your latest changes but I'm seeing this when I compile
![image](https://user-images.githubusercontent.com/82228511/178289330-0ebcdb0a-740c-45eb-b96b-77225a6b6487.png)
Is this something you've run into? Can you check if you need to merge development into your branch again (or rebase?)

pollend commented:
I don't think I've run into this ...

hultonha commented:
> Hey @pollend, I just tried building this again after grabbing your latest changes but I'm seeing this when I compile ![image](https://user-images.githubusercontent.com/82228511/178289330-0ebcdb0a-740c-45eb-b96b-77225a6b6487.png) Is this something you've run into? Can you check if you need to merge development into your branch again (or rebase?)

Tagging @burelc-amzn if they know what might have caused this (see above image)

rgba16f commented:
@hultonha There is a known issue where spheres are rendered inside out.
https://github.com/o3de/o3de/issues/6773

hultonha commented:
> @hultonha There is a known issue where spheres are rendered inside out. #6773

Aaah interesting! :D Thanks for the link @rgba16f, much appreciated 🙂

pollend commented:
@hultonha for number 9 I added it to that populateImpl but it doesn't seem to do anything

pollend commented on Gems/WhiteBox/Code/Source/SubComponentModes/EditorWhiteBoxTransformMode.cpp:
```diff
@@ -6,50 +6,640 @@
+                    {
+                        AzToolsFramework::ViewportUi::ViewportUiRequestBus::Event(
+                            AzToolsFramework::ViewportUi::DefaultViewportId,
+                            static_cast<void(AzToolsFramework::ViewportUi::ViewportUiRequestBus::Events::*)(
+                                    AzToolsFramework::ViewportUi::ClusterId, 
+                                    AzToolsFramework::ViewportUi::ButtonId)>(
+                                &AzToolsFramework::ViewportUi::ViewportUiRequestBus::Events::PressButton),
```
@hultonha you need the static cast since there are multiple implementations of PressButton.

```
        //! Invoke a button press on a cluster.
        virtual void PressButton(ClusterId clusterId, ButtonId buttonId) = 0;
        //! Invoke a button press on a switcher.
        virtual void PressButton(SwitcherId switcherId, ButtonId buttonId) = 0;
```

hultonha replied:
Aaah interesting 🤔 There actually might be a nicer way to do this using a lambda, I know @lumberyard-employee-dm often recommends this, might make the calling syntax a bit cleaner

lumberyard-employee-dm replied:
Yes that can be done by supplying a lambda whose first parameters is of the type of the EBus Interface pointer
```suggestion
                            [](AzToolsFramework::ViewportUi::ViewportUiRequests* requests)
                            {
                                requests->PressButton();
                            };
```

hultonha replied:
That was it! 😅 Thanks for adding the suggestion @lumberyard-employee-dm 👍

pollend replied:
umm kind of questionable if its that much shorter

pollend commented:
https://user-images.githubusercontent.com/854359/179389088-e83080db-cb9b-4837-8f53-b3561aa58cbb.mp4

pollend commented:
> @pollend unfortunately in your capture you mouse cursor isn't showing up, would it be possible to capture that too to make it easier to see what's going on. Can you let me know which items have been addressed too? We can potentially look at getting this PR merged as sooner rather than later (as it's off by default) but we will need to start addressing these bugs soon after if we do.


- I fixed number 6
- when clicking off the mesh I clear the selection number 7
- I make a distinction between hovering the cursor on the manipulator and the mesh so you can't highlight both the manipulator and the whitebox mesh.
- I added number 9

hultonha commented:
> > @pollend unfortunately in your capture you mouse cursor isn't showing up, would it be possible to capture that too to make it easier to see what's going on. Can you let me know which items have been addressed too? We can potentially look at getting this PR merged as sooner rather than later (as it's off by default) but we will need to start addressing these bugs soon after if we do.
> 
>     * I fixed number 6
> 
>     * when clicking off the mesh I clear the selection number 7
> 
>     * I make a distinction between hovering the cursor on the manipulator and the mesh so you can't highlight both the manipulator and the whitebox mesh.
> 
>     * I added number 9

Thanks very much for the update @pollend! I won't have time today but I'll grab the latest updates either tomorrow or Thursday and give some additional feedback. I'll take another look at the code then too. I do think fixing `3.` and `11.` would be worth fixing too as they make using the tool very difficult. The others we could then try and approach in a follow-up PR.

pollend commented:
> > > @pollend unfortunately in your capture you mouse cursor isn't showing up, would it be possible to capture that too to make it easier to see what's going on. Can you let me know which items have been addressed too? We can potentially look at getting this PR merged as sooner rather than later (as it's off by default) but we will need to start addressing these bugs soon after if we do.
> > 
> >     * I fixed number 6
> > 
> >     * when clicking off the mesh I clear the selection number 7
> > 
> >     * I make a distinction between hovering the cursor on the manipulator and the mesh so you can't highlight both the manipulator and the whitebox mesh.
> > 
> >     * I added number 9
> 
> Thanks very much for the update @pollend! I won't have time today but I'll grab the latest updates either tomorrow or Thursday and give some additional feedback. I'll take another look at the code then too. I do think fixing `3.` and `11.` would be worth fixing too as they make using the tool very difficult. The others we could then try and approach in a follow-up PR.

11 looks unrelated to this change. Where do I get assets for 3

pollend commented:
![image](https://user-images.githubusercontent.com/854359/180120694-9f85fe5b-e421-43bd-bd6b-9de599868c41.png)
I picked this icon umm seems to kind of workish

lumberyard-employee-dm commented on Gems/WhiteBox/Code/Source/SubComponentModes/EditorWhiteBoxTransformMode.cpp:
```diff
@@ -6,50 +6,667 @@
+    TransformMode::TransformMode(const AZ::EntityComponentIdPair& entityComponentIdPair)
+        : m_entityComponentIdPair(entityComponentIdPair)
+    {
+        AzToolsFramework::ViewportUi::ViewportUiRequestBus::EventResult(
+            m_transformClusterId,
+            AzToolsFramework::ViewportUi::DefaultViewportId,
+            &AzToolsFramework::ViewportUi::ViewportUiRequestBus::Events::CreateCluster,
+            AzToolsFramework::ViewportUi::Alignment::TopLeft);
+        m_transformTranslateButtonId = RegisterClusterButton(m_transformClusterId, "Move");
+        m_transformRotateButtonId = RegisterClusterButton(m_transformClusterId, "Rotate");
+        m_transformScaleButtonId = RegisterClusterButton(m_transformClusterId, "Scale");
+
+        // set translation tooltips
+        AzToolsFramework::ViewportUi::ViewportUiRequestBus::Event(
+            AzToolsFramework::ViewportUi::DefaultViewportId,
+            &AzToolsFramework::ViewportUi::ViewportUiRequestBus::Events::SetClusterButtonTooltip,
+            m_transformClusterId,
+            m_transformTranslateButtonId,
+            ManipulatorModeClusterTranslateTooltip);
+        AzToolsFramework::ViewportUi::ViewportUiRequestBus::Event(
+            AzToolsFramework::ViewportUi::DefaultViewportId,
+            &AzToolsFramework::ViewportUi::ViewportUiRequestBus::Events::SetClusterButtonTooltip,
+            m_transformClusterId,
+            m_transformRotateButtonId,
+            ManipulatorModeClusterRotateTooltip);
+        AzToolsFramework::ViewportUi::ViewportUiRequestBus::Event(
+            AzToolsFramework::ViewportUi::DefaultViewportId,
+            &AzToolsFramework::ViewportUi::ViewportUiRequestBus::Events::SetClusterButtonTooltip,
+            m_transformClusterId,
+            m_transformScaleButtonId,
+            ManipulatorModeClusterScaleTooltip);
```
FYI you can chain multiple calls to the same Bus Handler by using a lambda, since you have access to the Interface pointer.

Plus it is a more natural syntax since it is just several function calls

```suggestion

        AzToolsFramework::ViewportUi::ViewportUiRequestBus::Event(AzToolsFramework::ViewportUi::DefaultViewportId,
        [*this](AzToolsFramework::ViewportUi::ViewportUiRequests* requests)
        {
            m_transformClusterId = requests->CreateCluster(AzToolsFramework::ViewportUi::Alignment::TopLeft);
            m_transformTranslateButtonId = RegisterClusterButton(m_transformClusterId, "Move");
            m_transformRotateButtonId = RegisterClusterButton(m_transformClusterId, "Rotate");
            m_transformScaleButtonId = RegisterClusterButton(m_transformClusterId, "Scale");

            requests->SetClusterButtonTooltip(m_transformClusterId, m_transformTranslateButtonId, ManipulatorModeClusterTranslateTooltip);
            requests->SetClusterButtonTooltip(m_transformClusterId, m_transformRotateButtonId, ManipulatorModeClusterRotateTooltip);
            requests->SetClusterButtonTooltip(m_transformClusterId, m_transformScaleButtonId, ManipulatorModeClusterScaleTooltip);
        });
```

lumberyard-employee-dm commented on Code/Framework/AzToolsFramework/AzToolsFramework/Manipulators/RotationManipulators.cpp:
```diff
@@ -158,4 +158,22 @@ namespace AzToolsFramework
         m_circleBoundWidth = circleBoundWidth;
     }
+
+    AZStd::vector<BaseManipulator*> RotationManipulators::getManipulators()
```
Per the Coding Standard the function name should be [upper camel case](https://github.com/o3de/sig-core/blob/main/governance/Coding-Standards-and-Style-Guide.md#function-names).

Also Side Note: I am working on change to add support for C++[23 range conversion constructors](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2022/p1206r7.pdf) and utility functions which would make this code a little bit simpler
i.e
```c++
auto UniquePtrToRawPtr = [](auto uniquePtr) { return uniquePtr.get(); };
AZStd::vector<BaseManipulator*> manipulators(AZStd::from_range, m_localAngularManipulators | AZStd::transform(UniquePtrToRawPtr));
manipulators.append_range(m_viewAngularManipulator | AZStd::transform(UniquePtrToRawPtr));
return manipulators;
```

hultonha commented on Code/Framework/AzToolsFramework/AzToolsFramework/Manipulators/BaseManipulator.h:
```diff
@@ -300,6 +301,7 @@ namespace AzToolsFramework
         void SetLocalPosition(const AZ::Vector3& localPosition);
         void SetLocalOrientation(const AZ::Quaternion& localOrientation);
         void SetNonUniformScale(const AZ::Vector3& nonUniformScale);
+        virtual AZStd::vector<BaseManipulator*> GetManipulators() = 0;
```
Also this type doesn't really have any virtual functions (other than the slightly awkward `RefreshView` which some need), so it would be great if we don't have to add another and can solve this in a different way

hultonha commented on Code/Framework/AzToolsFramework/AzToolsFramework/Manipulators/RotationManipulators.cpp:
```diff
@@ -158,4 +158,22 @@ namespace AzToolsFramework
         m_circleBoundWidth = circleBoundWidth;
     }
+
+    AZStd::vector<BaseManipulator*> RotationManipulators::GetManipulators()
+    {
+        AZStd::vector<BaseManipulator*> manipualators;
+        for (auto& manipulator : m_localAngularManipulators)
+        {
+            if (manipulator)
+            {
+                manipualators.emplace_back(manipulator.get());
+            }
+        }
+        if (m_viewAngularManipulator)
+        {
+            manipualators.emplace_back(m_viewAngularManipulator.get());
+        }
+        return manipualators;
+    }
+
```
update: Adding this might not be the best way to achieve what we want, there could be an alternative design that does not require this. If you could give some more context on why this is needed and what you're trying to achieve I'd be very grateful

lumberyard-employee-dm replied:
There is a typo in the "manipualators" variable. It should be "manipulators".

Normally when I need to expose a container to client code, instead of returning a `vector`, `set` or some other container type, I instead prefer to provide a visitor function with a callback that the client can use to populate their own container

Such as 
```c++
using ManipulatorVisitCallback = AZStd::function<void(const AZStd::shared_ptr<AngularManipulator>&)>;
// ...
void RotationManipulators::VisitManipulators(const ManipulatorVisitCallback& callback) const
{
    for (const auto& manipulator : m_localAngularManipulators)
    {
        if (manipulator)
        {
           callback(manipulator);
        }
    }
    if (m_viewAngularManipulator)
    {
        callback(m_viewAngularManipulator);
    }
}
```

That way if you the client wanted to populate a vector or any other container type they can just a pass in a lambda.

```c++
AZStd::vector<BaseManipulator*> baseManipulators;
auto GetBaseManipulators = [&baseManipulator](const AZStd::shared_ptr<AngularManipulator>& angularManipulator)
{
    baseManipulators.push_back(angularManipulator.get());
};
rotationManipulators.VisitManipulators(GetBaseManipulators);
// Use the baseManipulators array as needed.
```

hultonha commented on Code/Framework/AzToolsFramework/AzToolsFramework/Manipulators/RotationManipulators.h:
```diff
@@ -44,6 +44,7 @@ namespace AzToolsFramework
 
         //! Sets the bound width to use for the circle (torus) of an angular manipulator.
         void SetCircleBoundWidth(float circleBoundWidth);
+        virtual AZStd::vector<BaseManipulator*> GetManipulators() override;
```
review/discuss: See earlier comments

hultonha commented on Code/Framework/AzToolsFramework/AzToolsFramework/Manipulators/ScaleManipulators.cpp:
```diff
@@ -157,4 +157,21 @@ namespace AzToolsFramework
         m_lineBoundWidth = lineBoundWidth;
     }
+
+    AZStd::vector<BaseManipulator*> ScaleManipulators::GetManipulators()
+    {
+        AZStd::vector<BaseManipulator*> manipulators;
+        for (auto& mainpulator : m_axisScaleManipulators)
+        {
+            if (mainpulator)
+            {
+                manipulators.emplace_back(mainpulator.get());
+            }
+        }
+        if (m_uniformScaleManipulator)
+        {
+            manipulators.emplace_back(m_uniformScaleManipulator.get());
+        }
+        return manipulators;
+    }
```
review/discuss: See earlier comments

hultonha commented on Code/Framework/AzToolsFramework/AzToolsFramework/Manipulators/ScaleManipulators.h:
```diff
@@ -43,6 +43,7 @@ namespace AzToolsFramework
 
         //! Sets the bound width to use for the line/axis of a linear manipulator.
         void SetLineBoundWidth(float lineBoundWidth);
+        virtual AZStd::vector<BaseManipulator*> GetManipulators() override;
```
review/discuss: See earlier comments

hultonha commented on Code/Framework/AzToolsFramework/AzToolsFramework/Manipulators/TranslationManipulators.cpp:
```diff
@@ -238,6 +238,29 @@ namespace AzToolsFramework
             m_surfaceManipulator->SetNonUniformScale(nonUniformScale);
         }
     }
+    AZStd::vector<BaseManipulator*> TranslationManipulators::GetManipulators()
+    {
+        AZStd::vector<BaseManipulator*> manipulators;
+        for (auto& mainpulator : m_linearManipulators)
+        {
+            if (mainpulator)
+            {
+                manipulators.emplace_back(mainpulator.get());
+            }
+        }
+        for (auto& mainpulator : m_planarManipulators)
+        {
+            if (mainpulator)
+            {
+                manipulators.emplace_back(mainpulator.get());
+            }
+        }
+        if (m_surfaceManipulator)
+        {
+            manipulators.emplace_back(m_surfaceManipulator.get());
+        }
+        return manipulators;
+    }
```
review/discuss: See earlier comments

hultonha commented on Code/Framework/AzToolsFramework/AzToolsFramework/Manipulators/TranslationManipulators.h:
```diff
@@ -76,6 +76,7 @@ namespace AzToolsFramework
 
         //! Sets the bound width to use for the line/axis of a linear manipulator.
         void SetLineBoundWidth(float lineBoundWidth);
+        virtual AZStd::vector<BaseManipulator*> GetManipulators() override;
```
review/discuss: See earlier comments

hultonha commented on Gems/WhiteBox/Code/Source/EditorWhiteBoxComponentMode.cpp:
```diff
@@ -542,7 +504,7 @@ namespace WhiteBox
             if (hasTransformMode)
             {
-                m_transformModeButtonId = RegisterClusterButton(m_modeSelectionClusterId, "TransformMode");
+                m_transformModeButtonId = RegisterClusterButton(m_modeSelectionClusterId, "Align_to_Object");
```
question: Why have we updated this from `TransformMode` to `"Align_to_Object", the previous name seemed better and more accurate

pollend replied:
might of copied it from somewhere

pollend replied:
its a temporary icon

pollend replied:
![image](https://user-images.githubusercontent.com/854359/182008788-1ae17e66-0628-4f8c-b3fa-d6f02574afd9.png)

hultonha replied:
Aaah I see! Sorry about that, thanks for clarifying 👍 (might be worth asking @amzn-leenguy for a new one based on the functionality)

hultonha replied:
Could you leave a comment to explain this just temporarily as that would be useful to understand why it's there, thanks!

hultonha commented on Gems/WhiteBox/Code/Source/EditorWhiteBoxComponentMode.h:
```diff
@@ -107,6 +101,7 @@ namespace WhiteBox
             m_keyboardMofifierQueryFn;
 
         SubMode m_currentSubMode = SubMode::Default;
+
```
nit: Might as well remove this new linebreak

hultonha commented on Gems/WhiteBox/Code/Source/SubComponentModes/EditorWhiteBoxTransformMode.h:
```diff
@@ -52,12 +76,47 @@ namespace WhiteBox
             const AZStd::optional<VertexIntersection>& vertexIntersection);
+
+    private:
+        struct VertexTransformSelection
```
suggestion: An API comment for this would be super helpful

hultonha commented on Gems/WhiteBox/Code/Source/SubComponentModes/EditorWhiteBoxTransformMode.h:
```diff
@@ -8,12 +8,25 @@
 #include "EditorWhiteBoxComponentModeTypes.h"
+#include "Viewport/WhiteBoxManipulatorViews.h"
+
+#include <AzCore/Math/Quaternion.h>
+#include <AzCore/std/containers/span.h>
+#include <AzCore/std/containers/variant.h>
+#include <AzCore/std/containers/vector.h>
+#include <AzCore/std/optional.h>
+#include <AzCore/std/smart_ptr/shared_ptr.h>
+#include <AzCore/std/smart_ptr/unique_ptr.h>
+#include <WhiteBox/WhiteBoxToolApi.h>
```
question: It does seem like some of these includes are either not required or could be moved to the .cpp file, might be worth doing a quick check. Thanks!

hultonha commented on Gems/WhiteBox/Code/Source/SubComponentModes/EditorWhiteBoxTransformMode.cpp:
```diff
@@ -6,50 +6,648 @@
 
-#include "EditorWhiteBoxComponentModeCommon.h"
 #include "EditorWhiteBoxTransformMode.h"
+#include "AzCore/std/smart_ptr/make_shared.h"
```
update: This include should be moved down with the `<>` style ones with `AzCore`

hultonha commented on Gems/WhiteBox/Code/Source/SubComponentModes/EditorWhiteBoxTransformMode.cpp:
```diff
@@ -6,50 +6,648 @@
+                .SetCallback(
+                    [cluserId = m_transformClusterId, buttonId = m_transformTranslateButtonId]()
+                    {
+                        
```
nit: Can we remove this line break please

hultonha commented on Gems/WhiteBox/Code/Source/SubComponentModes/EditorWhiteBoxTransformMode.cpp:
```diff
@@ -6,50 +6,648 @@
+        [[maybe_unused]] const AZ::EntityComponentIdPair& entityComponentIdPair)
     {
-        return {};
+        return {
```
Thanks for adding these shortcuts! 🙂 I'll check them out soon 👍

hultonha commented on Gems/WhiteBox/Code/Source/SubComponentModes/EditorWhiteBoxTransformMode.cpp:
```diff
@@ -6,50 +6,648 @@
+        EditorWhiteBoxComponentRequestBus::EventResult(
+            whiteBox, m_entityComponentIdPair, &EditorWhiteBoxComponentRequests::GetWhiteBoxMesh);
+
+        const bool mouseOverMainpulator = ([&]() 
```
nit: `mouseOverMainpulator` -> `mouseOverManipulator` (also better to be explicit with captures and not use a blanked `&`) (also there are extra parens around the expression which can be removed)

hultonha commented on Gems/WhiteBox/Code/Source/SubComponentModes/EditorWhiteBoxTransformMode.cpp:
```diff
@@ -6,50 +6,648 @@
+
+        const bool mouseOverMainpulator = ([&]() 
+            {
+                if(m_manipulator) 
+                {
+                    for(auto& manipulator: m_manipulator->GetManipulators()) 
+                    {
+                        if(manipulator->MouseOver()) 
+                        {
+                            return true;
+                        }
+                    }
+                }
+                return false;
```
update: I see this is why the `GetManipulators` was added, I do think there's another way to solve this, let me have a think and get back to you...

pollend replied:
you can just add a MouseOver() method? not sure if that is the ideal solution, don't think there is a situation where you only want to see if a particular part of the manipulator is hovered over.

hultonha replied:
Agreed, I think we can do something like this (discussed offline)

```
bool MouseOver()
{
    bool mouseOver = false;
    const auto mouseOverFn = [&mouseOver](BaseManipulator* manipulator)
    {
        mouseOver = manipulator->MouseOver() && mouseOver;
    };

    ProcessManipulators(mouseOverFn);

    return mouseOver;
}
```

Haven't tested this but I reckon something like this could work 🙂 (I hope! 😅)

hultonha commented on Gems/WhiteBox/Code/Source/SubComponentModes/EditorWhiteBoxTransformMode.cpp:
```diff
@@ -6,50 +6,648 @@
+        default:
+            break;
+        }
+        
```
nit: Please remove line break here

hultonha commented on Gems/WhiteBox/Code/Source/SubComponentModes/EditorWhiteBoxTransformMode.cpp:
```diff
@@ -6,50 +6,648 @@
+            AZStd::make_shared<AzToolsFramework::TranslationManipulators>(
+                AzToolsFramework::TranslationManipulators::Dimensions::Three, worldTranform, AZ::Vector3::CreateOne());
+
+        translationManipulators->SetLineBoundWidth(AzToolsFramework::ManipulatorLineBoundWidth());
```
Thanks for adding this 👍

hultonha reviewed (commented):
Some great updates, I just have a few questions outstanding. I think once the remaining questions are addressed and we can find an alternative to the `MouseOver` change we're good to go :+1: 🥳

hultonha commented on Code/Framework/AzToolsFramework/AzToolsFramework/Manipulators/RotationManipulators.cpp:
```diff
@@ -158,4 +158,20 @@ namespace AzToolsFramework
         m_circleBoundWidth = circleBoundWidth;
     }
+
+    void RotationManipulators::VisitManipulators(const ManipulatorVisitCallback&& callback)
+    {
+        for (auto& manipulator : m_localAngularManipulators)
+        {
+            if (manipulator)
+            {
+                callback(manipulator);
+            }
+        }
+        if (m_viewAngularManipulator)
+        {
+            callback(m_viewAngularManipulator);
+        }
+    }
+
```
I'm really sorry but we don't need this, it already exists in the form of `ProcessManipulators` (e.g `RotationManipulators::ProcessManipulators` and `ScaleManipulators::ProcessManipulators` etc.. - please remove these calls and take a look at my other comment

pollend replied:
do you want to make ProcessManipulators public?

pollend replied:
done

hultonha replied:
Perfect thanks! 👍

hultonha reviewed (commented):
I'm sorry but I really don't think we need to add `ManipulatorVisitCallback` and `VisitManipulators` - please see my most recent comment. I might be able to sync this branch and update with what I mean if you can hold off changes for a while. Thanks!

hultonha commented on Code/Framework/AzToolsFramework/AzToolsFramework/Manipulators/RotationManipulators.cpp:
```diff
@@ -158,4 +158,5 @@ namespace AzToolsFramework
     {
         m_circleBoundWidth = circleBoundWidth;
     }
+
```
uber nit: Can we please remove this white space change from this file

hultonha commented on Code/Framework/AzToolsFramework/AzToolsFramework/Manipulators/ScaleManipulators.h:
```diff
@@ -44,12 +44,12 @@ namespace AzToolsFramework
         //! Sets the bound width to use for the line/axis of a linear manipulator.
         void SetLineBoundWidth(float lineBoundWidth);
 
+        // Manipulators
```
nit: Please can we remove this comment, it's redundant here and adds nothing for the reader

hultonha commented on Gems/WhiteBox/Code/Source/EditorWhiteBoxComponentMode.cpp:
```diff
@@ -53,7 +64,7 @@ namespace WhiteBox
 
         // default behavior for querying modifier keys (ask the QApplication)
-        m_keyboardMofifierQueryFn = []()
+        m_keyboardModifierQueryFn = []()
```
Good spot! 😅 Thanks for fixing this 👍

hultonha commented on Gems/WhiteBox/Code/Source/EditorWhiteBoxComponentMode.h:
```diff
@@ -105,7 +99,7 @@ namespace WhiteBox
         AZ::Transform m_worldFromLocal;
         //! The function to use for querying modifier keys (while drawing).
         KeyboardModifierQueryFn
-            m_keyboardMofifierQueryFn;
+            m_keyboardModifierQueryFn;
```
suggestion/nit: Can we please move this variable onto the same line as the type while we're here

```
KeyboardModifierQueryFn m_keyboardModifierQueryFn;
```

hultonha commented on Gems/WhiteBox/Code/Source/SubComponentModes/EditorWhiteBoxTransformMode.cpp:
```diff
@@ -6,50 +6,636 @@
+            m_manipulator->ProcessManipulators(
+                [&mouseOverManipulator](auto manipulator)
+                {
+                        mouseOverManipulator = manipulator->MouseOver() || mouseOverManipulator;
```
uber nit: Looks like the indentation is off here (8 spaces instead of four) - suggest running clang-format before merging

hultonha reviewed (approved):
A few super minor suggestions that would be lovely to fix but otherwise is looking great 🙂 I'll review the change once it's merged next week (as we're still behind a feature flag for now). There'll likely be a few small things to tidy-up but getting closer to letting more customers try it out. Great work @pollend and thanks for your patience and determination!

amzn-leenguy commented:
Hi @pollend , I'm an UX designer interested in this area too. I just caught up on the thread. Incredible work so far @pollend ! Curious question, I see a blank icon in the Viewport UI cluster there. What is it supposed to be representing?

pollend commented:
> Hi @pollend , I'm an UX designer interested in this area too. I just caught up on the thread. Incredible work so far @pollend ! Curious question, I see a blank icon in the Viewport UI cluster there. What is it supposed to be representing? 

I need an icon 🤔

amzn-leenguy commented:
> > Hi @pollend , I'm an UX designer interested in this area too. I just caught up on the thread. Incredible work so far @pollend ! Curious question, I see a blank icon in the Viewport UI cluster there. What is it supposed to be representing?
> 
> I need an icon 🤔

Lol. can you help describe what the function is in one sentence. Just to make sure I am making the right thing.

pollend commented:
> > > Hi @pollend , I'm an UX designer interested in this area too. I just caught up on the thread. Incredible work so far @pollend ! Curious question, I see a blank icon in the Viewport UI cluster there. What is it supposed to be representing?
> > 
> > 
> > I need an icon thinking
> 
> Lol. can you help describe what the function is in one sentence. Just to make sure I am making the right thing.

its a translation,rotation, and scale mode. just lets the user do more fine tuning from the default mode of extruding and scaling.

hultonha commented:
Just kicked off a build for this @pollend :+1: 🙂

HogJonny-AMZN commented:
Really awesome to see these improvements coming in. Exciting.

hultonha commented:
Hey @pollend, unfortunately it looks like AR failed - https://jenkins.build.o3de.org/blue/rest/organizations/jenkins/pipelines/O3DE/branches/PR-9925/runs/2/nodes/472/log/?start=0. I'll try a clean build in case as it doesn't look super obvious but might be worth updating the branch. Thanks!

pollend commented:
> 

I merged back from develop. rebasing got a bit too messy to try.

hultonha commented:
> I merged back from develop. rebasing got a bit too messy to try.

Awesome thanks @pollend, I'll kick off another build now :+1:

pollend commented:
looks like there are some new errors ?

```
[2022-08-15T15:04:27.517Z] D:/workspace/o3de/Gems/WhiteBox/Code/Source/SubComponentModes/EditorWhiteBoxTransformMode.cpp(204,42): error C2220: the following warning is treated as an error [D:\workspace\o3de\build\windows\External\WhiteBox-285ae382\Code\WhiteBox.Editor.Static.vcxproj]

[2022-08-15T15:04:27.517Z] D:/workspace/o3de/Gems/WhiteBox/Code/Source/SubComponentModes/EditorWhiteBoxTransformMode.cpp(204,42): error C2220:         const IntersectionAndRenderData& renderData, [D:\workspace\o3de\build\windows\External\WhiteBox-285ae382\Code\WhiteBox.Editor.Static.vcxproj]

[2022-08-15T15:04:27.517Z] D:/workspace/o3de/Gems/WhiteBox/Code/Source/SubComponentModes/EditorWhiteBoxTransformMode.cpp(204,42): error C2220:                                          ^ [D:\workspace\o3de\build\windows\External\WhiteBox-285ae382\Code\WhiteBox.Editor.Static.vcxproj]

[2022-08-15T15:04:28.446Z]   EditorPythonBindings.Editor.vcxproj -> D:\workspace\o3de\build\windows\bin\profile\EditorPythonBindings.Editor.dll

[2022-08-15T15:04:29.826Z]   PrefabBuilder.Static.vcxproj -> D:\workspace\o3de\build\windows\lib\profile\PrefabBuilder.Static.lib

[2022-08-15T15:04:35.096Z] D:/workspace/o3de/Gems/WhiteBox/Code/Source/SubComponentModes/EditorWhiteBoxTransformMode.cpp(204,42): warning C4100: 'renderData': unreferenced formal parameter [D:\workspace\o3de\build\windows\External\WhiteBox-285ae382\Code\WhiteBox.Editor.Static.vcxproj]

[2022-08-15T15:04:35.106Z] D:/workspace/o3de/Gems/WhiteBox/Code/Source/SubComponentModes/EditorWhiteBoxTransformMode.cpp(204,42): warning C4100:         const IntersectionAndRenderData& renderData, [D:\workspace\o3de\build\windows\External\WhiteBox-285ae382\Code\WhiteBox.Editor.Static.vcxproj]

[2022-08-15T15:04:35.106Z] D:/workspace/o3de/Gems/WhiteBox/Code/Source/SubComponentModes/EditorWhiteBoxTransformMode.cpp(204,42): warning C4100:                                          ^ [D:\workspace\o3de\build\windows\External\WhiteBox-285ae382\Code\WhiteBox.Editor.Static.vcxproj]

```

hultonha commented:
@pollend Aah it's the dreaded unused variable warning, I'll kick off another build now 👍

pollend commented:
> @pollend Aah it's the dreaded unused variable warning, I'll kick off another build now +1

:D it passed

hultonha commented:
> > @pollend Aah it's the dreaded unused variable warning, I'll kick off another build now +1
> 
> :D it passed

Amazing! 😄 Can we get a US person (@lumberyard-employee-dm? 😁) to merge otherwise I can do this first thing tomorrow morning 😎

lumberyard-employee-dm commented:
> > > @pollend Aah it's the dreaded unused variable warning, I'll kick off another build now +1
> > 
> > 
> > :D it passed
> 
> Amazing! 😄 Can we get a US person (@lumberyard-employee-dm? 😁) to merge otherwise I can do this first thing tomorrow morning 😎

I can merge it.

hultonha commented:
Awesome thank you @lumberyard-employee-dm! 😄 Much appreciated 🙂

amzn-leenguy commented:
@hultonha @lumberyard-employee-dm Is this live in the nightly builds? I would love to play with it tomorrow. 9/2

lumberyard-employee-dm commented:
Yes, it should be available.

hultonha commented:
Correct it's behind a feature flag, you can enable it with `/O3DE/Preferences/WhiteBox/TransformFeature`, you can enable it from the Editor console with `--regset=/O3DE/Preferences/WhiteBox/TransformFeature=true`

pollend commented:
> @hultonha @lumberyard-employee-dm Is this live in the nightly builds? I would love to play with it tomorrow. 9/2

very cool, would like to hear your thoughts :P
